src/ametsuchi/merkle_tree/merkle_tree.cc

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <ametsuchi/merkle_tree/merkle_tree.h>
+
+namespace ametsuchi {
+namespace merkle {
+
+/**
+ * Returns floor(log2(x))
+ * log2(0) => undefined behaviour
+ */
+static inline size_t log2(size_t x) {
+  size_t y = 0;
+  while (x > 1) {
+    x >>= 1;
+    y++;
+  }
+  return y;
+}
+
+/**
+ * Returns integer log2 of x
+ * ceil2(5) = 8
+ */
+static inline size_t ceil2(size_t x) {
+  size_t l = 1u << log2(x);
+  return l == x ? l : l + 1;
+}
+
+MerkleTree::MerkleTree(size_t leafs) : leafs_(0) {
+  // round number of leafs to the power of 2
+  leafs_ = ceil2(leafs);
+
+  // full tree size
+  size_ = leafs_ * 2 - 1;
+  tree_.resize(size_);
+
+  i_current_ = leafs_ - 1;
+  i_root_ = i_current_;
+}
+
+hash_t MerkleTree::root() { return tree_[i_root_]; }
+
+void MerkleTree::push(const hash_t &item) {
+  if (i_current_ == leafs_ - 1) {
+    // this is the very first push. just move item to the leftmost leaf
+    tree_[i_current_] = item;
+    i_root_ = i_current_++;
+    return;
+  }
+
+  if (i_current_ == size_) {
+    // tree is complete, logically means creation of a NEW BLOCK
+    tree_[leafs_ - 1] = tree_[0];  // copy root to leftmost leaf
+    i_root_ = leafs_ - 1;          // change root pointer
+    i_current_ = leafs_;           // change pointer to current free cell
+  }
+
+  // copy hash to current empty position
+  tree_[i_current_] = item;
+
+  // find LCA(leftmost leaf, i_current_)
+  // LCA is this many levels above:
+  size_t np = 1 + log2(i_current_ - (leafs_ - 1));
```

Comment:
Also, yes, I will add a rollback feature soon. I have to complete generator of transactions.